DSESwift/DataSyncer.swift

```diff
@@ -8,37 +8,398 @@
 
 import Foundation
 import RealmSwift
+import PromiseKit
+import CoreLocation
+import SwiftyJSON
 
-class DataSyncer: NSObject {
 
-    func login() {
-    downloadSensorProfile ()
+
+/**
+ * DataSyncer handles the synchronization between the local storage and CommonSense.
+ * The syncing process is handled automatically and periodically, thus the external
+ * user does not need to be aware of the data syncing process at all.
+ *
+ */
+public class DataSyncer : NSObject {
+    
+    enum DataSyncerError: ErrorType{
+        case InvalidPersistentPeriod
+        case InvalidSyncRate
+    }
+    
+    static let SOURCE = "aim-ios-sdk"
+    let SENSOR_PROFILE_KEY_NAME = "sensor_name"
+    let SENSOR_PROFILE_KEY_STRUCTURE = "data_structure"
+    
+    let MAX_INITIALIZATION_ATTEMPTS = 5
+    
+    var initializationAttemps = 0
+
+    // only used for the default parameters
+    let config = DSEConfig()
+    
+    var syncRate: Double!
+    var persistentPeriod: Double!
+    
+    var timer:NSTimer?
+    let data_syncer_process_queue = dispatch_queue_create("nl.sense.dse.sync_process_queue", nil)
+    
+    var delegate: DataSyncerDelegate?
+    
+    private(set) var initialized = false
+
+    // the key for the string should be <source>:<sensor>
+    override init () {
+        self.syncRate = self.config.syncInterval!
+        self.persistentPeriod = self.config.localPersistancePeriod!
+    }
+    
+    /**
+     * this will error check it's own configuration
+     */
+    func setConfig(config: DSEConfig) throws -> (configChanged: Bool, syncInterval: Double, localPersistancePeriod: Double) {
+        var configChanged = false
+        if let syncInterval = config.syncInterval {
+            if (syncInterval < 0) {throw DataSyncerError.InvalidSyncRate}
+            configChanged = configChanged || self.syncRate != syncInterval
+            self.syncRate = syncInterval
+        }
+        
+        if let localPersistancePeriod = config.localPersistancePeriod {
+            if (localPersistancePeriod < 0) {throw DataSyncerError.InvalidPersistentPeriod}
+            configChanged = configChanged || self.persistentPeriod != localPersistancePeriod
+            self.persistentPeriod = localPersistancePeriod
+        }
+        
+        // return new values that will be used
+        return (configChanged, self.syncRate, self.persistentPeriod)
+    }
+    
+    func initialize()  {
+        dispatch_async(data_syncer_process_queue, {
+            do{
+                try self.downloadSensorProfiles()
+                print("DSE Initialization Completed")
```

Comment:
do we still print here? Is this not for debugging? Is there a difference between logging and printing in swift? Just wondering...